From cb306c1776d8909d70d92dd29626412bbedef14e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:03:49 -0400
Subject: [usb] usbtest: support container id descriptor test

Message-id: <1410887463-20674-158-git-send-email-dzickus@redhat.com>
Patchwork-id: 95211
O-Subject: [RHEL7 PATCH 157/591] usb: usbtest: support container id descriptor test
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 8e29217f931dd8a76cdb8c88aa0975fa15f0f684
Author: Huang Rui <ray.huang@amd.com>
Date:   Wed Oct 30 11:27:40 2013 +0800

    usb: usbtest: support container id descriptor test

    In Test 9 of usbtest module, it is used for performing chapter 9 tests N
    times.

    Container ID descriptor is one of the generic device-level capbility
    descriptors which added in section 9.6.2.3 of USB 3.0 spec.

    This patch adds to support getting Container ID descriptor test scenario
    for USB 3.0.

    Signed-off-by: Huang Rui <ray.huang@amd.com>
    Acked-by: Alan Stern <stern@rowland.harvard.edu>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/misc/usbtest.c b/drivers/usb/misc/usbtest.c
index d6fcbc4..b415282 100644
--- a/drivers/usb/misc/usbtest.c
+++ b/drivers/usb/misc/usbtest.c
@@ -657,6 +657,25 @@ static int is_good_ss_cap(struct usbtest_dev *tdev, u8 *buf)
  return 1;
 }
 
+static int is_good_con_id(struct usbtest_dev *tdev, u8 *buf)
+{
+ struct usb_ss_container_id_descriptor *con_id;
+
+ con_id = (struct usb_ss_container_id_descriptor *) buf;
+
+ if (con_id->bLength != USB_DT_USB_SS_CONTN_ID_SIZE) {
+  ERROR(tdev, "bogus container id descriptor length\n");
+  return 0;
+ }
+
+ if (con_id->bReserved) { /* reserved == 0 */
+  ERROR(tdev, "reserved bits set\n");
+  return 0;
+ }
+
+ return 1;
+}
+
 /* sanity test for standard requests working with usb_control_mesg() and some
  * of the utility functions which use it.
  *
@@ -810,6 +829,14 @@ static int ch9_postconfig(struct usbtest_dev *dev)
      return -EDOM;
     }
     break;
+   case CONTAINER_ID_TYPE:
+    if (buf + USB_DT_USB_SS_CONTN_ID_SIZE >
+      dev->buf + total ||
+      !is_good_con_id(dev, buf)) {
+     dev_err(&iface->dev, "bogus container id descriptor\n");
+     return -EDOM;
+    }
+    break;
    default:
     break;
    }
-- 
1.7.1